### PR TITLE
BugFix(#541): collect function list first, then replace usages with n…

### PR DESCRIPTION
**Pull request by sjw36**
…ew vectorized version

### CODE REVIEW

File: lib/SplatArgPass.cpp [hunk 1]
```diff
@@ -152,32 +153,31 @@ bool SplatArgPass::runOnModule(Module &M) {
...
+    auto &func_info = FI.second;
+    auto func_type = func_info.getType();
+    auto &param_info = func_info.getParameter(0);
+    bool has_3_params =
+        func_type == Builtins::kClamp || func_type == Builtins::kMix;
+    std::string NewFName = getSplatName(func_info, param_info, has_3_params);
+    Function *NewCallee = getReplacementFunction(*F, NewFName);
+    // Replace the users of the function.
+    for (User *U : F->users()) {
+      replaceCall(NewCallee, dyn_cast<CallInst>(U));
```
Comment (jrprice) [inline lib/SplatArgPass.cpp]: This is still modifying the user list while iterating over it, so my test case still hangs with this code.

**Comment by sjw36** [comment]
> FWIW `ninja check-spirv` passes for me even though the kernel I posted in #541 hangs, so I don't think we do have coverage for this specific issue.

Interested to know your build os/env @jrprice, thanks.

File: lib/SplatArgPass.cpp (hunk 1 above)
Comment (alan-baker) [reply lib/SplatArgPass.cpp]: Thanks for catching that.

@sjw36 I'd suggest using a SmallVector to copy the users before modifying them.

Comment (sjw36) [reply lib/SplatArgPass.cpp]: Too right, I will fix that..

Comment (sjw36) [reply lib/SplatArgPass.cpp]: I was just going to :
  while (!F->use_empty()) {
    User *U = *F->users().begin();
    replaceCall(...);
  }
Good?

**Comment by jrprice** [comment]
> Interested to know your build os/env @jrprice

Debian-based Linux distro, building with GCC 9.

File: lib/SplatArgPass.cpp (hunk 1 above)
Comment (alan-baker) [reply lib/SplatArgPass.cpp]: Looks ok, but I don't have time to try it out.

**Comment by sjw36** [comment]
Any ideas why this keeps failing?

CMake Error at CMakeLists.txt:1518 (add_custom_command):
  No TARGET 'libGLESv2' has been created in this directory.

**Comment by alan-baker** [comment]
Rebase on top of master. That has been fixed.

File: lib/SplatArgPass.cpp [hunk 2]
```diff
@@ -135,10 +135,11 @@ void SplatArgPass::replaceCall(Function *NewCallee, CallInst *Call) {
...
-  for (auto &F : M) {
+  std::list<std::pair<Function *, const Builtins::FunctionInfo &>> func_list;
+  // Collect candidates
+  for (auto &F : M.getFunctionList()) {
     // process only function declarations
-    if (F.empty() && !F.use_empty()) {
+    if (F.isDeclaration() && !F.use_empty()) {
       auto &func_info = Builtins::Lookup(&F);
       auto func_type = func_info.getType();
       switch (func_type) {
```
Comment (jrprice) [inline lib/SplatArgPass.cpp]: The `default` clause in this switch seems to have been dropped for some reason:
```warning: 289 enumeration values not handled in switch:```

Comment (sjw36) [reply lib/SplatArgPass.cpp]: Great, I will add the default clause back as well.

**Pull request merged**